crates/wasi/src/ctx.rs

```diff
@@ -325,18 +499,101 @@ impl WasiCtxBuilder {
+    /// [`preview1`]: crate::preview1
+    ///
+    /// # Panics
+    ///
+    /// Panics if this method is called twice. Each [`WasiCtxBuilder`] can be
+    /// used to create only a single [`WasiCtx`] or [`WasiP1Ctx`]. Repeated
+    /// usage of this method is not allowed and should use a second builder
+    /// instead.
     #[cfg(feature = "preview1")]
-    pub fn build_p1(&mut self) -> WasiP1Ctx {
+    #[cfg_attr(docsrs, doc(cfg(feature = "preview1")))]
+    pub fn build_p1(&mut self) -> crate::preview1::WasiP1Ctx {
         let wasi = self.build();
-        WasiP1Ctx::new(wasi)
+        crate::preview1::WasiP1Ctx::new(wasi)
     }
 }
 
+/// A trait which provides access to internal WASI state.
+///
+/// This trait is the basis of implementation of all traits in this crate. All
+/// traits are implemented like:
+///
+/// ```
+/// # trait WasiView {}
+/// # mod bindings { pub mod wasi { pub trait Host {} } }
+/// impl<T: WasiView> bindings::wasi::Host for T {
+///     // ...
+/// }
+/// ```
+///
+/// For a [`Store<T>`](wasmtime::Store) this trait will be implemented
+/// for the `T`. This also corresponds to the `T` in
+/// [`Linker<T>`](wasmtime::component::Linker).
+///
+/// # Example
+///
+/// ```
+/// use wasmtime_wasi::{WasiCtx, ResourceTable, WasiView, WasiCtxBuilder};
+///
+/// struct MyState {
+///     ctx: WasiCtx,
+///     table: ResourceTable,
+/// }
+///
+/// impl WasiView for MyState {
+///     fn ctx(&mut self) -> &mut WasiCtx { &mut self.ctx }
+///     fn table(&mut self) -> &mut ResourceTable { &mut self.table }
+/// }
+///
+/// impl MyState {
+///     fn new() -> MyState {
+///         let mut wasi = WasiCtxBuilder::new();
+///         wasi.arg("./foo.wasm");
+///         wasi.arg("--help");
+///         wasi.env("FOO", "bar");
+///
+///         MyState {
+///             ctx: wasi.build(),
+///             table: ResourceTable::new(),
+///         }
+///     }
+/// }
+/// ```
 pub trait WasiView: Send {
+    /// Yields mutable access to the internal resource management that this
+    /// context contains.
+    ///
+    /// Embedders can add custom resources to this table as well to give
+    /// resources to wasm as well.
     fn table(&mut self) -> &mut ResourceTable;
+
+    /// Yields mutable access to the configuration used for this context.
+    ///
+    /// The returned type is created through [`WasiCtxBuilder`].
     fn ctx(&mut self) -> &mut WasiCtx;
 }
 
+/// Per-[`Store`] state which holds state necessary to implement WASI from this
+/// crate.
+///
+/// This structure is created through [`WasiCtxBuilder`] and is stored within
+/// the `T` of [`Store<T>`][`Store`]. Access to the structure is provided
+/// through the [`WasiView`] trait as an implementation on `T`.
+///
+/// Note that this structure itself does not have any accessors, it's here for
+/// internal use within the `wasmtime-wasi` crate's implementation of
+/// bindgen-generated traits.
+///
+/// [`Store`]: wasmtime::Store
 pub struct WasiCtx {
```

Comment:
Not related to this PR exactly, but one pattern that can be very useful when a builder is required to build a type is to a `builder` function to that type. So in this case:

```rust
WasiCtx::builder();
```